### PR TITLE
Update JITServer documentation

### PR DESCRIPTION
**Pull request by dmitry-ten**
- Remove/fix outdated information: option names, protobuf mentions
- Update usage and testing guides
- Add more detailed documentation for resolved methods
- Add tips for debugging JITServer

### CODE REVIEW

**Comment by dmitry-ten** [comment]
@mpirvu fyi.
@EmanElsaban when you have time, I'd like your opinion on this as well.
I tried to incorporate most of the points you made. Adding documentation for Jenkins testing would be nice, but personal builds are on internal jenkins, so I don't think we can add that. I could add info for running adopt builds, but I'm not sure if there is already a guide for that.

File: doc/compiler/jitserver/Problem.md [hunk 1]
```diff
@@ -19,9 +19,40 @@ OpenJDK Assembly Exception [2].
...
+## Client-side crashes
 If the client crashes in `handleServerMessage` after receiving some bad data from the server, you will need to find out what the server was doing when it sent the message. The easiest way to do this is by running both the server and client in gdb, then sending the server a Ctrl-C after the client crashes. From the gdb prompt on the server you can type `thread apply all backtrace` and find the appropriate compilation thread to determine where the server was.
 
-Please add to this file!!
+## Crashes in compiled code
+Another way a client can crash is when it runs the compiled method but something is wrong with it. These can be some of the tougher bugs to fix
+because you will need to find the failing method and look at its trace log to find the problem. Sometimes, the failure is not caused by one compilation
+but by a series of them, which makes debugging even harder.
+
+Tracing method compilation is the same as for non-JITServer - use `-Xjit:{<method_name>}(traceFull,log=<log_name>)` in the client options and both client
```
Comment (mpirvu) [inline doc/compiler/jitserver/Problem.md:45]: Can we add somewhere in here about the double compilation debugging feature? This is the one where a method is compiled both at the server and at the client for the purpose of comparing the two compilation logs.

File: doc/compiler/jitserver/Problem.md [hunk 2]
```diff
@@ -19,9 +19,40 @@ OpenJDK Assembly Exception [2].
 
 SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 -->
+# Debugging JITServer
 
-Usually typical OpenJ9 debugging procedures can be followed for JITServer. But there are a few JITServer specific tips and tricks.
+Typical OpenJ9 debugging procedures can be followed for JITServer. But there are a few JITServer specific tips and tricks.
+Unlike regular OpenJ9, where everything runs in one process, JITServer separates JIT compiler into a separate process while constantly
```
Comment (mpirvu) [inline doc/compiler/jitserver/Problem.md:25]: "Unlike regular OpenJ9 where everything runs in one process, JITServer technology moves the JIT compiler activity into a separate process while ..."

File: doc/compiler/jitserver/ResolvedMethod.md [hunk 3]
```diff
@@ -20,8 +20,52 @@ OpenJDK Assembly Exception [2].
...
+or having incorrect value cached for the duration of one compilation does not have impact on correctness or performance (i.e. whether a method is interpreted).
+
+Resolved methods could also be used as more general per-compilation caches, containing information that does not directly describe the resolved method itself.
+Since resolved methods are destroyed at the end of compilation, this would be functionally correct.
+However, such usage is discouraged because `TR::CompilationInfoPerThreadRemote` is a better place for such caches. Placing caches there makes them more visible
+and it makes more sense to store information not related to resolved methods outside of them.
+
+### Performance considerations
+
+Since resolved methods are so frequently created and used, the number of remote messages associated with them is normally a large portion of total messages. This means that minimizing the number of requests for resolved method creation is important for improving CPU consumption.
```
Comment (mpirvu) [inline doc/compiler/jitserver/ResolvedMethod.md:57]: "This means that minimizing" --> "Therefore, minimizing..."

File: doc/compiler/jitserver/ResolvedMethod.md [hunk 4]
```diff
@@ -20,8 +20,52 @@ OpenJDK Assembly Exception [2].
...
+Whenever a resolved method needs to be created, server will first check the cache and recreate the resolved method from there, if possible.
+2. Prefetching multiple resolved methods before they are first requested. There are parts of the compiler that request many resolved methods in a loop.
+For example, ILGen and inliner will walk through method bytecodes and create a resolved method for every method call bytecode. 
+We use `TR_ResolvedJ9JITServerMethod::cacheResolvedMethodsCallees` to do a preliminary walk and cache all required methods in one remote call.
+Techniques like these do not reduce the total number of resolved methods created or data transferred through the network but it reduces the total number of remote calls,
+which still has a positive impact on CPU consumption.
+
+Despite all these optimizations, resolved methods are still responsible for most of the remote messages. This is due to the fact that resolved methods only exist
+for the duration of a single compilation so any type of persistent caching is not available to us. Maybe there exists some way to get around this limitation
```
Comment (mpirvu) [inline doc/compiler/jitserver/ResolvedMethod.md:69]: Please delete the last two sentences: "Maybe there exists..."

File: doc/compiler/jitserver/Usage.md [hunk 5]
```diff
@@ -53,23 +53,37 @@ With verbose logging, if a client connects successfully then server output shoul
...
+To check client's version pass it `verbose=\{JITServer\}` option as well.
+```
+$ java -XX:+UseJITServer -Xjit:verbose=\{JITServer\} -version
+
+JITServer is currently a technology preview. Its use is not yet supported
+#JITServer: JITServer version: 1.17.0
+#JITServer: JITServer Client Mode. Server address: localhost port: 38400. Connection Timeout 2000ms
+#JITServer: Identifier for current client JVM: 9238414254742342824
 ...
 ```
```
Comment (mpirvu) [inline doc/compiler/jitserver/Usage.md:89]: Maybe we should also mention about the heart-beat?

**Review by mpirvu — COMMENTED** [review]
I have some small comments. Also, all headers need to change the year to 2021 now.

File: doc/compiler/jitserver/Problem.md (hunk 1 above)
Comment (dmitry-ten) [reply doc/compiler/jitserver/Problem.md:45]: Right, I actually forgot that this option exists.

**Comment by dmitry-ten** [comment]
@mpirvu updated.

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
Changes are for documentation only, so no testing is required.

**Pull request merged**